.devops/deploy-pipelines.yml

```diff
@@ -0,0 +1,151 @@
+          ),
+          endsWith(variables['Build.SourceBranch'], '-TEST-PIPELINE')
+        )
+      )
+    jobs:
+      - job: make_release
+        steps:
+        - ${{ if eq(variables['Build.SourceBranch'], 'refs/heads/master') }}:
+          - template: templates/node-github-release/template.yaml@pagopaCommons 
+            parameters:
+              semver: '${{ parameters.RELEASE_SEMVER }}'
+              gitEmail: $(GIT_EMAIL)
+              gitUsername: $(GIT_USERNAME)
+              gitHubConnection: $(GITHUB_CONNECTION)
+              nodeVersion: $(NODE_VERSION)
+              pkg_cache_version_id: $(CACHE_VERSION_ID)
+              pkg_cache_folder: $(YARN_CACHE_FOLDER)
+              release_branch: $(gitBranchName)
+
+        - ${{ if ne(variables['Build.SourceBranch'], 'refs/heads/master') }}:
+          - script: |
+              echo "We assume this reference to be a valid release: $(Build.SourceBranch). Therefore, there is no need to bundle a new release."
+            displayName: 'Skip release bundle'
+
+  # Prepare Artifact
+  - stage: Prepare_artifact
+    dependsOn:
+      - Release
+    jobs:
+      - job: make_build
+        steps:
+          - template: azure-templates/setup-project.yml
+            parameters:
+              # On the assumption that this stage is executed only when Relase stage is,
+              #  with this parameter we set the reference the deploy script must pull changes from.
+              # The branch/tag name is calculated from the source branch
+              #  ex: Build.SourceBranch=refs/heads/master --> master
+              #  ex: Build.SourceBranch=refs/tags/v1.2.3-RELEASE --> v1.2.3-RELEASE
+              gitReference: ${{ replace(replace(variables['Build.SourceBranch'], 'refs/tags/', ''), 'refs/heads/', '') }}
+
+          - script: |
+              yarn build
+            displayName: 'Build'
+
+          - task: CopyFiles@2
+            inputs:
+              SourceFolder: '$(System.DefaultWorkingDirectory)'
+              TargetFolder: '$(System.DefaultWorkingDirectory)/bundle'
+              Contents: |
+                **/*
+                !.git/**/*
+                !**/*.js.map
+                !**/*.ts
+                !src/**/*
+                !.vscode/**/*
+                !.devops/**/*
+                !.prettierrc
+                !.gitignore
+                !jest.config.js
+                !test
+                !tsconfig.json
+                !tslint.json
+                !yarn.lock
+                !Dangerfile.js
+                !CODEOWNERS
+                !__*/**/*
+            displayName: 'Copy bundle files'
+          
+          - publish: $(System.DefaultWorkingDirectory)/bundle
+            artifact: Bundle
+
+  # Publish
+  # skip this step for -TEST-PIPELINE branches
+  - stage: Publish
+    dependsOn:
+      - Prepare_artifact
+    condition:
+    not(
+        endsWith(variables['Build.SourceBranch'], '-TEST-PIPELINE')
+    )
```

Comment:
I like the idea very much, I'm just a bit unsure about the execution.

This condition refers to the whole stage, but what we actually want to avoid is just the npm publish task. 
I think we should run the download step and then choose to run the publish task or not. 

In this way, we can push the test further.

Also, I'm concerned about using  convention over the branch name. What's the user scenario?

Reply:
I don't know if it's gonna work, it's just a silly idea to try.
The user scenario is exactly this one, so when we need to test the outcome of the pipeline itself.
As I said in my first comment:

I tried to add the possibility to test the deploy pipeline into a dedicated branch (named -TEST-PIPELINE)
In this case it should skip the process of publishing the artifact to npm, and it should create the changelog into the branch itself instead of master.
This feature needs to be tested and verified.

In this case, for testing the deploy without publishing we need to switch to a new branch, with a proper suffix.
Not a problem to change the suffix, do you have any suggestion? (meanwhile I'm gonna create a variable for it)

Reply:
> I like the idea very much, I'm just a bit unsure about the execution.
> 
> This condition refers to the whole stage, but what we actually want to avoid is just the npm publish task.
> I think we should run the download step and then choose to run the publish task or not.

What do you mean? I thought that the publish step was made in the _Prepare_artifact_ stage, and published as artifact. I was waiting to test the possibility of downloading it from Azure directly, but if there is a better way, happy to hear because I don't like this step neither.

Reply:
I like the attitude of questioning "how am I going to test it?" <3

I'd rather not use a convention on branch name, instead we can use an optional variable like `DRY_RUN`, which we can set on true when we want the pipeline to have no side effects.

Please consider that the package publishing is not the only side effect we have - we also edit the repository. To apply changes in a disposable branch might be a solution, but once you push a tag, such tag belongs to the repository not only to a single branch; hence we should avoid pushing tags.

We can of course review the whole process for supporting a dry run, but I'd rather implement a thoughtful solution.

What do you think?

PS: If you want to skip a whole stages, you can disable them just before activating the pipeline. That's useful for a quick check like "let's pretend I run the pipeline with this configuration".

Reply:
> I'd rather not use a convention on branch name, instead we can use an optional variable like `DRY_RUN`, which we can set on true when we want the pipeline to have no side effects.

Love this solution, I was thinking if it was possible!

>Please consider that the package publishing is not the only side effect we have - we also edit the repository. To apply changes in a disposable branch might be a solution, but once you push a tag, such tag belongs to the repository not only to a single branch; hence we should avoid pushing tags.

I tried to consider it by setting the **release_branch** parameter of **templates/node-github-release/template.yaml@pagopaCommons**

Reply:
>  having a "test environment" for pipelines could be useful for more impacting applications.

That is absolutely a good point, we should have a test environment for pipelines. 

Me and @pasqualedevita were thinking about design a solution for that, let's move this conversation somewhere else.

About this PR, I would not introduce a test layer for now

Reply:
fine for me, I'll remove all the related code..